File: app/src/main/java/com/aslansari/hypocoin/ui/adapters/BaseAdapter.java
package com.aslansari.hypocoin.ui.adapters;

import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import timber.log.Timber;

public abstract class BaseAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    // region Constants
    protected static final int HEADER = 0;
    protected static final int ITEM = 1;
    protected static final int FOOTER = 2;
    // endregion

    // region Member Variables
    protected List<T> items;
    protected OnItemClickListener onItemClickListener;
    protected OnReloadClickListener onReloadClickListener;
    protected OnLongClickListener onLongClickListener;
    protected boolean isFooterAdded = false;
    // endregion

    // region Constructors
    public BaseAdapter() {
        items = new ArrayList<>();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder = null;

        switch (viewType) {
            case HEADER:
                viewHolder = createHeaderViewHolder(parent);
                break;
            case ITEM:
                viewHolder = createItemViewHolder(parent);
                break;
            case FOOTER:
                viewHolder = createFooterViewHolder(parent);
                break;
            default:
                break;
        }

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int position) {
        switch (getItemViewType(position)) {
            case HEADER:
                bindHeaderViewHolder(viewHolder);
                break;
            case ITEM:
                bindItemViewHolder(viewHolder, position);
                break;
            case FOOTER:
                bindFooterViewHolder(viewHolder);
            default:
                break;
        }
    }

    @Override
    public int getItemCount() {
        return items.size();
    }
    // endregion

    // region Abstract Methods
    protected abstract RecyclerView.ViewHolder createHeaderViewHolder(ViewGroup parent);
    // endregion

    protected abstract RecyclerView.ViewHolder createItemViewHolder(ViewGroup parent);

    protected abstract RecyclerView.ViewHolder createFooterViewHolder(ViewGroup parent);

    protected abstract void bindHeaderViewHolder(RecyclerView.ViewHolder viewHolder);

    protected abstract void bindItemViewHolder(RecyclerView.ViewHolder viewHolder, int position);

    protected abstract void bindFooterViewHolder(RecyclerView.ViewHolder viewHolder);

    protected abstract void displayLoadMoreFooter();

    protected abstract void displayErrorFooter();

    public abstract void addFooter();

    // region Helper Methods
    public T getItem(int position) {
        return items.get(position);
    }

    public void add(T item) {
        items.add(item);
        notifyItemInserted(items.size() - 1);
    }

    public void add(int position, T item) {
        items.add(position, item);
        notifyItemInserted(position);
    }

    public void addAll(List<T> items) {
        for (T item : items) {
            add(item);
        }
    }
    // endregion

    public void addAll(int position, List<T> items) {
        for (T item : items) {
            add(position, item);
        }
    }

    public void remove(T item) {
        int position = items.indexOf(item);
        if (position > -1) {
            items.remove(position);
            notifyItemRemoved(position);
        }
    }

    public void clear() {
        Timber.d("clear recyclerView");
        isFooterAdded = false;
        while (getItemCount() > 0) {
            remove(getItem(0));
        }
    }

    public boolean isEmpty() {
        return getItemCount() == 0;
    }

    public boolean isLastPosition(int position) {
        return (position == items.size() - 1);
    }

    public boolean isFooterAdded() {
        return isFooterAdded;
    }

    public void removeFooter() {
        isFooterAdded = false;

        int position = items.size() - 1;
        T item = getItem(position);

        if (item != null) {
            items.remove(position);
            notifyItemRemoved(position);
        }
    }

    public void updateFooter(FooterType footerType) {
        switch (footerType) {
            case LOAD_MORE:
                displayLoadMoreFooter();
                break;
            case ERROR:
                displayErrorFooter();
                break;
            default:
                break;
        }
    }

    public void setOnLongClickListener(OnLongClickListener onLongClickListener) {
        this.onLongClickListener = onLongClickListener;
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public void setOnReloadClickListener(OnReloadClickListener onReloadClickListener) {
        this.onReloadClickListener = onReloadClickListener;
    }

    // region Inner Classes
    public enum FooterType {
        LOAD_MORE,
        ERROR
    }

    // region Interfaces
    public interface OnItemClickListener {
        void onItemClick(int position, View view);
    }

    public interface OnReloadClickListener {
        void onReloadClick();
    }

    public interface OnLongClickListener {
        void onItemLongClick(int position, View view);
    }
    // endregion

    public interface ISortable {
        void sort();
    }
    // endregion
}

File: app/src/main/java/com/aslansari/hypocoin/ui/adapters/BaseAdapter.kt
package com.aslansari.hypocoin.ui.adapters

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import timber.log.Timber

abstract class BaseAdapter<T> : RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
    // endregion
    // region Member Variables
    protected val HEADER = 0
    protected val ITEM = 1
    protected val FOOTER = 2
    @JvmField
    protected var items: MutableList<T>
    protected var onItemClickListener: OnItemClickListener? = null
        set
    protected var onReloadClickListener: OnReloadClickListener? = null
        set
    protected var onLongClickListener: OnLongClickListener? = null
        set
    var isFooterAdded = false
        protected set

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        var viewHolder: RecyclerView.ViewHolder? = null
        when (viewType) {
            HEADER -> viewHolder = createHeaderViewHolder(parent)
            ITEM -> viewHolder = createItemViewHolder(parent)
            FOOTER -> viewHolder = createFooterViewHolder(parent)
            else -> {}
        }
        return viewHolder!!
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            HEADER -> bindHeaderViewHolder(holder)
            ITEM -> bindItemViewHolder(holder, position)
            FOOTER -> bindFooterViewHolder(holder)
            else -> {}
        }
    }

    override fun getItemCount(): Int {
        return items.size
    }

    // endregion
    // region Abstract Methods
    protected abstract fun createHeaderViewHolder(parent: ViewGroup?): RecyclerView.ViewHolder?

    // endregion
    protected abstract fun createItemViewHolder(parent: ViewGroup?): RecyclerView.ViewHolder?
    protected abstract fun createFooterViewHolder(parent: ViewGroup?): RecyclerView.ViewHolder?
    protected abstract fun bindHeaderViewHolder(viewHolder: RecyclerView.ViewHolder?)
    protected abstract fun bindItemViewHolder(viewHolder: RecyclerView.ViewHolder?, position: Int)
    protected abstract fun bindFooterViewHolder(viewHolder: RecyclerView.ViewHolder?)
    protected abstract fun displayLoadMoreFooter()
    protected abstract fun displayErrorFooter()
    abstract fun addFooter()

    // region Helper Methods
    fun getItem(position: Int): T {
        return items[position]
    }

    fun add(item: T) {
        items.add(item)
        notifyItemInserted(items.size - 1)
    }

    fun add(position: Int, item: T) {
        items.add(position, item)
        notifyItemInserted(position)
    }

    fun addAll(items: List<T>) {
        for (item in items) {
            add(item)
        }
    }

    // endregion
    fun addAll(position: Int, items: List<T>) {
        for (item in items) {
            add(position, item)
        }
    }

    fun remove(item: T) {
        val position = items.indexOf(item)
        if (position > -1) {
            items.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    fun clear() {
        Timber.d("clear recyclerView")
        isFooterAdded = false
        while (itemCount > 0) {
            remove(getItem(0))
        }
    }

    val isEmpty: Boolean
        get() = itemCount == 0

    fun isLastPosition(position: Int): Boolean {
        return position == items.size - 1
    }

    fun removeFooter() {
        isFooterAdded = false
        val position = items.size - 1
        val item: T? = getItem(position)
        if (item != null) {
            items.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    fun updateFooter(footerType: FooterType?) {
        when (footerType) {
            FooterType.LOAD_MORE -> displayLoadMoreFooter()
            FooterType.ERROR -> displayErrorFooter()
            else -> {}
        }
    }



    // region Inner Classes
    enum class FooterType {
        LOAD_MORE, ERROR
    }

    // region Interfaces
    interface OnItemClickListener {
        fun onItemClick(position: Int, view: View?)
    }

    interface OnReloadClickListener {
        fun onReloadClick()
    }

    interface OnLongClickListener {
        fun onItemLongClick(position: Int, view: View?)
    }

    // endregion
    interface ISortable {
        fun sort()
    } // endregion

    companion object {
        // region Constants
    }

    // endregion
    // region Constructors
    init {
        items = ArrayList()
    }
}